### PR TITLE
added a new keyboard focused JSON editor

### PR DESCRIPTION
**Pull request by ArnoldSmith86**
Yesterday evening I had a few ideas and was thinking: what would the editor look like that I would want to use?

It is inspired by the discussion we had last meeting about #135. This is meant for power users who edit JSON manually on the PC with a big screen. It will **definitely not** replace any of the other editors (well, the JSON one maybe).

I have **a lot** more ideas and plans for this but now is the first time where it might be useful.

The basic ideas:

- JSON editing front and center. Complete browser height to have as much visible as possible.
- **Many** context-sensitive hotkeys using `CTRL` and `SHIFT` to modify the JSON and the widgets. Most of them visible on the screen.
- The room itself should always be visible and playable.

Press `CTRL+J` to toggle it. Then `CTRL`-click on a widget. I'll add a hint for that somewhere appropriate.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3176/pr-test (or any other room on that server)

**Comment by larkob** [comment]
When I press Ctrl+J, I just get a blank screen on the right hand side and the sidebar on the left disappears.

**Comment by ArnoldSmith86** [comment]
Whoops, forgot this part: Then `CTRL`-click on a widget.

**Comment by larkob** [comment]
Alright. Quite a good start and helps to quickly add properties. The shortcuts change everytime I add a property, what's the idea here? Should shortcuts like Ctrl+H always add specific properties (like `height`)? Currently, it seems completely random. Every time I press Ctrl+H, it adds some other property. I know it's early alpha, so I'm just interested if you've got some specific plans for the shortcut mapping yet.

**Comment by ArnoldSmith86** [comment]
Will have another look later but it tries the letters in the property in order and uses the first free one. If none are free, it goes through the alphabet. Unfortunately there are more blocked once than I initially thought (n, q, w, t for tab management; x, c, v for clipboard).

The list should also be sorted by command name. I'm still right in the middle of this but I thought I'd share the version that might actually be useful already.

**Comment by RaphaelAlvez** [comment]
Why does the comands keep changing? Is this normal in programming editors?
I guess that pressing crtl+m twice is practical but it's a little confusing to me.

**Comment by robartsd** [comment]
Love the zoom out to show off-screen elements. Can interact with off-screen elements directly, but if you move them you're limited to dropping them in the room. Should be great for working with off-screen button routines.

This may quickly become my favorite editor.

**Comment by robartsd** [comment]
Would love to have code folding. With code folding, would be great to be able to see entire room state (start with all widgets collapsed).

**Comment by robartsd** [comment]
A way to easily navigate from card to deck could be very helpful. Can get deck in edit mode but not in play mode. No way to toggle edit/play mode while editor is open.

**Comment by robartsd** [comment]
A way to add a new empty widget would also be great.

**Comment by ArnoldSmith86** [comment]
> Would love to have code folding. With code folding, would be great to be able to see entire room state (start with all widgets collapsed).

Stuff like code folding would require quite a lot of stuff.

**Comment by robartsd** [comment]
Perhaps just a way to open a widget by ID?

**Comment by rogerl50** [comment]
This is great. Undo would be really useful, too.

**Comment by robartsd** [comment]
New organization of the command pane looks great!

**Comment by robartsd** [comment]
Adjustable pane width?

I'd probably just do this with user style once in main if you choose not to implement. On my 1080p screen 15% seems about right for the command pane. Haven't figured out how to resize the room pane yet (but 35% for the text editor isn't bad either).

**Comment by ArnoldSmith86** [comment]
I'll probably set a fixed width and let the room take the rest. I'm adding stuff above and below the room though and I'll wait for that.

File: client/js/jsonedit.js [hunk 1]
```diff
@@ -34,6 +34,13 @@ const jeCommands = [
       return widgets.has(m[1]);
     }
   },
+  {
+    name: 'upload a different asset',
+    context: '"(/assets/[0-9_-]+)"',
```
Comment (robartsd) [inline client/js/jsonedit.js:39]: Not sure what you could do to improve this functionality, but if the user misses any character at the end of the asset URL the results after upload will be broken.

Also a way to add assets where none has been uploaded yet would be nice. (I can type a dummy asset URL to replace).

Even as it is now, this is a huge improvement to work in JSON with uploaded assets.

**Comment by RaphaelAlvez** [comment]
My suggestions. Be able to see two widgets json.

Have fixed comands for every scenario. For example if you are in general properties then ctrl+n would always add movableInEdit but Ctrl+n would still be available in other situations like click routines.

And make it in night mode and add some of those nice coding colors

**Comment by robartsd** [comment]
Updating a widget unparent children.

**Comment by ArnoldSmith86** [comment]
> Ctrl-Fx should replace null with quoted string.

Yes.

It ignores `movableInEdit:false` atm. I will probably change that.

**Comment by ArnoldSmith86** [comment]
I still have **a lot** of ideas for this.

To the people currently actually making games: What are missing features that would save you the most time now?

**Comment by RaphaelAlvez** [comment]
Defnetly two tabs for me. Get information from other widgets while building a routine.

**Comment by robartsd** [comment]
> I still have **a lot** of ideas for this.
> 
> To the people currently actually making games: What are missing features that would save you the most time now?

Not sure, but I think it's already the best editor available for working in JSON.

**Comment by RaphaelAlvez** [comment]
How about a console to text routines?

Maybe give it a test mode where it simulates the routines and then comes back and a run mode where it just applies the routine

**Comment by ArnoldSmith86** [comment]
I added the ability to write Javascript macros to the editor. Those are executed with something similar to `eval`. I feel like that's not a problem because it's only executed on the machine where it's entered. You can basically do the same in the browser dev tools anyway.

Does anyone know a reason why it's a bad idea?

**Comment by rogerl50** [comment]
> I added the ability to write Javascript macros to the editor. Those are executed with something similar to `eval`. I feel like that's not a problem because it's only executed on the machine where it's entered. You can basically do the same in the browser dev tools anyway.
> 
> Does anyone know a reason why it's a bad idea?

Why not just add the built-in macros from ghetto while you're at it? I assume those are executed the same way.

**Comment by ArnoldSmith86** [comment]
I still have many ideas for this but I might slow down quite a bit. So maybe merge it as it is (unless there are serious bugs) and then add features with issues and PRs?

**Comment by 96LawDawg** [comment]
Yes.  Perfection is the enemy of good enough.  I would rather have what there is now than wait for it to get incrementally better.

**Comment by rogerl50** [comment]
> I still have many ideas for this but I might slow down quite a bit. So maybe merge it as it is (unless there are serious bugs) and then add features with issues and PRs?

Agreed. Go for it.

**Review by 96LawDawg — APPROVED** [review]
I have not reviewed the code, but I have tested the features.  I have also seen the features demonstrated.  There are still a few crashes and strange glitches, but it works good enough and is ready for version 1.

**Review by RaphaelAlvez — APPROVED** [review]
Awesome feature

**Pull request merged**